File: tests/test_pps.py
"""PPS 系列单测（与 C++ tests/unit/test_pps.cpp 断言一致）。"""

import sys
import os

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from googology.notations.ordinal.sequence.marked_parent.pps import (
    Pps1, Pps2, Pps4, WPps4, TPps4, FPps4,
)
from googology.core import NotComparable


def check(cond, msg):
    if not cond:
        raise AssertionError(msg)


def expect_throw(fn, exc, msg):
    try:
        fn()
    except exc:
        return
    raise AssertionError(msg)


# ---- PPS1 展开：坏根 = 第 x 项（x = 末项值）----
check(Pps1("(0, 1, 0, 3)").to_latex() == "(0, 1, 0, 3)", "parse (0,1,0,3)")
check(Pps1("0,1,0,3").to_latex() == "(0, 1, 0, 3)", "parse bare")
check(Pps1("[0,1,0,3]").to_latex() == "(0, 1, 0, 3)", "parse []")

p = Pps1("(0,1,2)"); p.expand(0)
check(p.to_latex() == "(0, 1, 1)", "pps1 (0,1,2).expand(0)")
p = Pps1("(0,1,2)"); p.expand(1)
check(p.to_latex() == "(0, 1, 1, 1)", "pps1 (0,1,2).expand(1)")
p = Pps1("(0,1,2)"); p.expand(2)
check(p.to_latex() == "(0, 1, 1, 1, 1)", "pps1 (0,1,2).expand(2)")

p = Pps1("(0,1,0,3)"); p.expand(0)
check(p.to_latex() == "(0, 1, 0, 2)", "pps1 (0,1,0,3).expand(0)")
p = Pps1("(0,1,0,3)"); p.expand(1)
check(p.to_latex() == "(0, 1, 0, 2, 2)", "pps1 (0,1,0,3).expand(1)")

p = Pps1("(0,1,0,1,2)"); p.expand(0)
check(p.to_latex() == "(0, 1, 0, 1, 1, 0, 1)", "pps1 (0,1,0,1,2).expand(0)")

# 后继：末项为 0 → 去掉末项
p = Pps1("(0,1,0)"); p.expand(3)
check(p.to_latex() == "(0, 1)", "pps1 (0,1,0).expand(3)")
check(Pps1("(0,1,0)").is_successor, "pps1 (0,1,0).is_successor")
check(not Pps1("(0,1,2)").is_successor, "pps1 (0,1,2) not successor")

# expand_to = expandLen
p = Pps1("(0,1,2)"); p.expand_to(2)
check(p.to_latex() == "(0, 1, 1, 1, 1)", "pps1 (0,1,2).expand_to(2)")

# 坏根越界 → ValueError
expect_throw(lambda: Pps1("(0,5)").expand(0), ValueError, "pps1 (0,5) bad-root OOR")

# ---- PPS1 vs PPS2 ----
p = Pps1("(0,2,4,4,2,3)"); p.expand(0)
check(p.to_latex() == "(0, 2, 4, 4, 2, 4, 7, 2)", "pps1 (0,2,4,4,2,3).expand(0)")
p = Pps2("(0,2,4,4,2,3)"); p.expand(0)
check(p.to_latex() == "(0, 2, 4, 4, 2, 2, 7, 2)", "pps2 (0,2,4,4,2,3).expand(0)")
p = Pps2("(0,1,2)"); p.expand(0)
check(p.to_latex() == "(0, 1, 1)", "pps2 (0,1,2).expand(0)")

# ---- PPS4 弱/强展开 ----
p = Pps4("(0,1,0,1,2)"); p.expand(0)
check(p.to_latex() == "(0, 1, 0, 1, 1, 0, 1)", "pps4 weak expand(0)")
p = Pps4("(0,1,0,2,2,3)"); p.expand(0)
check(p.to_latex() == "(0, 1, 0, 2, 2, 1, 2, 2)", "pps4 strong col expand(0)")
p = Pps1("(0,1,0,2,2,3)"); p.expand(0)
check(p.to_latex() == "(0, 1, 0, 2, 2, 2, 2, 2)", "pps1 same input")
p = Pps4("(0,1,2,1,3)"); p.expand(0)
check(p.to_latex() == "(0, 1, 2, 1, 2, 1)", "pps4 strong not-found")

# ---- wPPS4（与 PPS4 一致）----
p = WPps4("(0,1,0,2,2,3)"); p.expand(0)
check(p.to_latex() == "(0, 1, 0, 2, 2, 1, 2, 2)", "wpps4 expand(0)")

# ---- tPPS4 / fPPS4 强展开复制修饰 ----
p = TPps4("(0,1,0,2,2,3)"); p.expand(1)
check(p.to_latex() == "(0, 1, 0, 2, 2, 1, 2, 2, 4, 2, 2)", "tpps4 expand(1)")
p = Pps4("(0,1,0,2,2,3)"); p.expand(1)
check(p.to_latex() == "(0, 1, 0, 2, 2, 1, 2, 2, 1, 2, 2)", "pps4 expand(1)")
p = FPps4("(0,1,0,2,2,3)"); p.expand(1)
check(p.to_latex() == "(0, 1, 0, 2, 2, 1, 2, 2, 3, 2, 2)", "fpps4 expand(1)")

# ---- §12 limit API ----
check(Pps1.limit(0).to_latex() == "()", "limit(0)")
check(Pps1.limit(3).to_latex() == "(0, 1, 2)", "limit(3)")
ml = Pps1.master_limit(); ml.expand(3)
check(ml.to_latex() == "(0, 1, 2)", "master_limit.expand(3)")

# ---- compare ----
check(Pps1("(0,1,2)").compare(Pps1("(0,1,3)")) == -1, "compare <")
check(Pps1("(0,1,3)").compare(Pps1("(0,1,2)")) == 1, "compare >")
check(Pps1("(0,1,2)").compare(Pps1("(0,1,2)")) == 0, "compare ==")
check(Pps1("(0,1,2)").compare(Pps1("(0,1,2,0)")) == -1, "compare prefix")
expect_throw(lambda: Pps1("(0,1,2)").compare(Pps2("(0,1,2)")), NotComparable,
             "cross-variant compare")
from googology.notations.number.knuth import Knuth
expect_throw(lambda: Pps1("(0,1,2)").compare(Knuth("2 ^ 3")), NotComparable,
             "cross-family compare")

# ---- __getitem__（非修改）----
check(Pps1("(0,1,2)")[1].to_latex() == "(0, 1, 1, 1)", "A[1]")

# ---- taxonomy / capability ----
p = Pps1()
check(p.name == "pps1" and Pps2().name == "pps2" and Pps4().name == "pps4"
      and WPps4().name == "wpps4" and TPps4().name == "tpps4"
      and FPps4().name == "fpps4", "variant names")
check(p.style == "marked_parent", "style")
check(p.creator == "318`4", "creator")
check(p.version == "1" and TPps4().version == "4t", "versions")
check(p.can(__import__("googology.core", fromlist=["Op"]).Op.Compare), "can Compare")

print("test_pps: PASS")

File: tests/test_pps.c
// PPS 系列 C 分支单测（与 C++ tests/unit/test_pps.cpp 断言一致）。
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "googology/notations/ordinal/sequence/marked_parent/pps/pps.h"

static void expect_latex(goog_notation_t* n, const char* expected) {
    char* s = goog_to_latex(n);
    assert(strcmp(s, expected) == 0);
    free(s);
}

int main(void) {
    // ---- parse / to_string ----
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS1, "(0, 1, 0, 3)");
        expect_latex(p, "(0, 1, 0, 3)");
        goog_destroy(p);
    }
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS1, "0,1,0,3");
        expect_latex(p, "(0, 1, 0, 3)");
        goog_destroy(p);
    }

    // ---- PPS1 展开：坏根 = 第 x 项 ----
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS1, "(0,1,2)");
        goog_expand(p, 0);
        expect_latex(p, "(0, 1, 1)");
        goog_destroy(p);
    }
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS1, "(0,1,2)");
        goog_expand(p, 1);
        expect_latex(p, "(0, 1, 1, 1)");
        goog_destroy(p);
    }
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS1, "(0,1,2)");
        goog_expand(p, 2);
        expect_latex(p, "(0, 1, 1, 1, 1)");
        goog_destroy(p);
    }
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS1, "(0,1,0,3)");
        goog_expand(p, 0);
        expect_latex(p, "(0, 1, 0, 2)");
        goog_destroy(p);
    }
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS1, "(0,1,0,3)");
        goog_expand(p, 1);
        expect_latex(p, "(0, 1, 0, 2, 2)");
        goog_destroy(p);
    }
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS1, "(0,1,0,1,2)");
        goog_expand(p, 0);
        expect_latex(p, "(0, 1, 0, 1, 1, 0, 1)");
        goog_destroy(p);
    }

    // ---- 后继：末项为 0 → 去掉末项 ----
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS1, "(0,1,0)");
        goog_expand(p, 3);
        expect_latex(p, "(0, 1)");
        goog_destroy(p);
    }

    // ---- expand_to = expandLen ----
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS1, "(0,1,2)");
        goog_expand_to(p, 2);
        expect_latex(p, "(0, 1, 1, 1, 1)");
        goog_destroy(p);
    }

    // ---- 坏根越界：保持原样（C++ throws，此处 panic-free）----
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS1, "(0,5)");
        goog_expand(p, 0);
        expect_latex(p, "(0, 5)");
        goog_destroy(p);
    }

    // ---- PPS1 vs PPS2 ----
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS1, "(0,2,4,4,2,3)");
        goog_expand(p, 0);
        expect_latex(p, "(0, 2, 4, 4, 2, 4, 7, 2)");
        goog_destroy(p);
    }
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS2, "(0,2,4,4,2,3)");
        goog_expand(p, 0);
        expect_latex(p, "(0, 2, 4, 4, 2, 2, 7, 2)");
        goog_destroy(p);
    }

    // ---- PPS4 弱/强展开 ----
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS4, "(0,1,0,2,2,3)");
        goog_expand(p, 0);
        expect_latex(p, "(0, 1, 0, 2, 2, 1, 2, 2)");
        goog_destroy(p);
    }
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS1, "(0,1,0,2,2,3)");
        goog_expand(p, 0);
        expect_latex(p, "(0, 1, 0, 2, 2, 2, 2, 2)");
        goog_destroy(p);
    }
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS4, "(0,1,2,1,3)");
        goog_expand(p, 0);
        expect_latex(p, "(0, 1, 2, 1, 2, 1)");
        goog_destroy(p);
    }

    // ---- wPPS4（与 PPS4 一致）----
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_WPPS4, "(0,1,0,2,2,3)");
        goog_expand(p, 0);
        expect_latex(p, "(0, 1, 0, 2, 2, 1, 2, 2)");
        goog_destroy(p);
    }

    // ---- tPPS4 / fPPS4 强展开复制修饰 ----
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_TPPS4, "(0,1,0,2,2,3)");
        goog_expand(p, 1);
        expect_latex(p, "(0, 1, 0, 2, 2, 1, 2, 2, 4, 2, 2)");
        goog_destroy(p);
    }
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS4, "(0,1,0,2,2,3)");
        goog_expand(p, 1);
        expect_latex(p, "(0, 1, 0, 2, 2, 1, 2, 2, 1, 2, 2)");
        goog_destroy(p);
    }
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_FPPS4, "(0,1,0,2,2,3)");
        goog_expand(p, 1);
        expect_latex(p, "(0, 1, 0, 2, 2, 1, 2, 2, 3, 2, 2)");
        goog_destroy(p);
    }

    // ---- §12 limit API ----
    {
        goog_notation_t* p = goog_pps_limit(PPS_VARIANT_PPS1, 3);
        expect_latex(p, "(0, 1, 2)");
        goog_destroy(p);
    }
    {
        goog_notation_t* p = goog_pps_master_limit(PPS_VARIANT_PPS1);
        goog_expand(p, 3);
        expect_latex(p, "(0, 1, 2)");
        goog_destroy(p);
    }

    // ---- compare：字典序，跨变体 NotComparable ----
    {
        goog_notation_t* a = goog_pps_parse(PPS_VARIANT_PPS1, "(0,1,2)");
        goog_notation_t* b = goog_pps_parse(PPS_VARIANT_PPS1, "(0,1,3)");
        assert(a->vtable->compare(a, b) == -1);
        assert(b->vtable->compare(b, a) == 1);
        assert(a->vtable->compare(a, a) == 0);
        goog_notation_t* c = goog_pps_parse(PPS_VARIANT_PPS1, "(0,1,2,0)");
        assert(a->vtable->compare(a, c) == -1);
        goog_destroy(a); goog_destroy(b); goog_destroy(c);
    }
    {
        goog_notation_t* a = goog_pps_parse(PPS_VARIANT_PPS1, "(0,1,2)");
        goog_notation_t* b = goog_pps_parse(PPS_VARIANT_PPS2, "(0,1,2)");
        assert(a->vtable->compare(a, b) == GOOG_ERR_NOT_COMPARABLE);
        goog_destroy(a); goog_destroy(b);
    }

    // ---- taxonomy ----
    {
        goog_notation_t* p = goog_pps_parse(PPS_VARIANT_PPS1, "(0,1,2)");
        assert(strcmp(goog_name(p), "pps1") == 0);
        assert(goog_family(p) == FAMILY_ORDINAL);
        assert(strcmp(p->vtable->creator(p), "318`4") == 0);
        assert(strcmp(p->vtable->version(p), "1") == 0);
        assert(goog_can(p, OP_COMPARE));
        assert(goog_can(p, OP_EXPAND));
        goog_destroy(p);
        goog_notation_t* t = goog_pps_parse(PPS_VARIANT_TPPS4, "(0,1,2)");
        assert(strcmp(goog_name(t), "tpps4") == 0);
        assert(strcmp(t->vtable->version(t), "4t") == 0);
        goog_destroy(t);
    }

    printf("test_pps: PASS\n");
    return 0;
}
